File: DesignPatterns/Creational/AbstractFactory/ComputerPartsClient.java
interface CPU {
  public void displayCPU();
}

class EmberCPU implements CPU {

  public void displayCPU(){
    System.out.println("Ember CPU is been built");
  }
}

class EnginolaCPU implements CPU {

  public void displayCPU() {
    System.out.println("Enginola CPU is been build");
  }
}

interface MMU {
  public void displayMMU();
}

class EmberMMU implements MMU {

  public void displayMMU(){
    System.out.println("Ember MMU is been built");
  }
}

class EnginolaMMU implements MMU {

  public void displayMMU() {
    System.out.println("Enginola MMU is been built");
  }
}

enum Architecture {
  ENGINOLA, EMBER
}

abstract class ToolKitFactory {
  private static final EmberToolKit EMBER_TOOLKIT = new EmberToolKit();
  private static final EnginolaToolKit ENGINOLA_TOOLKIT = new EnginolaToolKit();

  static ToolKitFactory getFactory(Architecture architecture) {
    ToolKitFactory factory = null;
    switch(architecture) {
      case EMBER:
        factory = EMBER_TOOLKIT;
        break;
      case ENGINOLA:
        factory = ENGINOLA_TOOLKIT;
        break;
    }
    return factory;
  }

  public abstract CPU createCPU();
  public abstract MMU createMMU();
}

class EmberToolKit extends ToolKitFactory {

  public CPU createCPU(){
    return new EmberCPU();
  }

  public MMU createMMU() {
    return new EmberMMU();
  }
}

class EnginolaToolKit extends ToolKitFactory {

  public CPU createCPU() {
    return new EnginolaCPU();
  }

  public MMU createMMU() {
    return new EnginolaMMU();
  }
}

public class ComputerPartsClient {
  public static void main(String[] args) {
    ToolKitFactory factory = ToolKitFactory.getFactory(Architecture.EMBER);
    CPU cpu = factory.createCPU();
    cpu.displayCPU();
  }
}

File: DesignPatterns/Creational/AbstractFactory/Ruby/computer_parts_client.rb
module Constants
  ENGINOLA = "ENGINOLA".freeze
  EMBER = "EMBER".freeze
end

class CPU
  def display_cpu
    raise NotImplementedError
  end
end

class EmberCPU < CPU
  def display_cpu
    puts 'Ember CPU is been built'
  end
end

class EnginolaCPU < CPU
  def display_cpu
    puts 'Enginola CPU is been build'
  end
end

class MMU
  def display_mmu
    raise NotImplementedError
  end
end

class EmberMMU < MMU
  def display_mmu
    puts 'Ember MMU is been built'
  end
end

class EnginolaMMU < MMU
  def display_mmu
    puts 'Enginola MMU is been built'
  end
end

class AbstractFactory
  def create_cpu
    raise NotImplementedError
  end

  def create_mmu
    raise NotImplementedError
  end
end

class EmberToolkit < AbstractFactory
  def create_cpu
    EmberCPU.new
  end

  def create_mmu
    EmberMMU.new
  end
end

class EnginolaToolkit < AbstractFactory
  def create_cpu
    EnginolaCPU.new
  end

  def create_mmu
    EnginolaMMU.new
  end
end


class ToolKitFactory
  @@ember_toolkit = EmberToolkit.new
  @@enginola_tookit = EnginolaToolkit.new

  def self.fetch_factory(architecture)
    case architecture
    when Constants::EMBER then @@ember_toolkit
    when Constants::ENGINOLA then @@enginola_tookit
    end
  end
end

tf = ToolKitFactory.fetch_factory(Constants::EMBER)
tf.create_cpu.display_cpu
tf.create_mmu.display_mmu